binary_search_tree/tree.py

```diff
@@ -14,42 +14,129 @@ class Tree:
+        return current_root
+
+    # Time Complexity:  O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+
+        if self.root is None:
+            self.root = new_node
+            return
+        self.add_helper(self.root, new_node)
+        
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def find_helper(self, current, key):
+        if current == None:
+            return None
+        elif current.key == key:
+            return current.value
+        elif current.key > key:
+            return self.find_helper(current.left, key)
+        elif current.key < key:
+            return self.find_helper(current.right, key)
+
+    # Time Complexity: O(log n) -- Assuming it's balanced
+    # Space Complexity: O(log n)
     def find(self, key):
-        pass
+        if  self.root == None:
+            return None
+        return self.find_helper(self.root, key)
+        
+
+    def inorder_helper(self, current, items):
+        if current != None:
+            self.inorder_helper(current.left, items)
+            items.append({"key": current.key, "value": current.value})
+            self.inorder_helper(current.right, items)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+        self.inorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def preorder_helper(self, current, items):
+        if current != None:
+            items.append({"key": current.key, "value": current.value})
+            self.preorder_helper(current.left, items)
+            self.preorder_helper(current.right, items)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def postorder(self):
-        pass
+    def postorder_helper(self, current, items):
+        if current != None:
+            self.postorder_helper(current.left, items)
+            self.postorder_helper(current.right, items)
+            items.append({"key": current.key, "value": current.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+    # Time Complexity: O(n) 
+    # Space Complexity: O(n)  
+    def postorder(self):
+        items = []
+        self.postorder_helper(self.root, items)
+        return items
 
+    def height_helper(self, current):
+        if current != None:
+            height = max(self.height_helper(current.left), self.height_helper(current.right))
+            return height + 1
+        else: 
+            return 0
 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
+    def height(self):
```

Comment:
👍  The space complexity is correct in the worst case if the tree can be unbalanced and O(log n) if the tree is balanced.